src/utils/inferencerUtils.ts

```diff
@@ -75,32 +75,48 @@ const predefined = new Set([
   'undefined'
 ])
 
-export function printTypeConstraints(typeContraints: Map<number, number | string>) {
+function printType(type: Type): string {
+  switch (type.kind) {
+    case "primitive":
+      return type.name
+    case "variable":
+      return `T${type.id}`
+    case "function":
+      let params = ""
+      for (const argument of type.parameterTypes) {
+        params += printType(argument) + ", "
+      }
+      // remove last comma
+      params = params.replace(/,\s*$/, "");
+      const returnType = printType(type.returnType)
+      return `(${params}) => ${returnType}`
+    default:
+      return "Not included in Source 1!"
+  }
+}
+
+export function printTypeConstraints(typeContraints: Map<Type, Type>) {
   console.log('Printing Type Constraints')
   for (const [key, value] of typeContraints) {
-    console.log(`T${key} = T${value}`)
+    console.log(`${printType(key)} = ${printType(value)}`)
   }
   console.log("\n");
 }
 
-export function printTypeEnvironment(typeEnvironment: Map<any, any>) {
+export function printTypeEnvironment(typeEnvironment: Map<string, any>) {
   console.log('Printing Type Environment')
   for (let [key, value] of typeEnvironment) {
     if (predefined.has(key)) {
       continue
     }
-    if (typeof value === 'object') {
-      value = JSON.stringify(value)
-    }
-    console.log(`${key} = T${value}`)
+    console.log(`${key} = ${printType(value.types[0])}`)
```

Comment:
`${printType(value.types[0])}`)` - I think this part will give error. 

The current TypeEnv is `...types: [{argumentTypes: [Type, Type], resultType: Type}]...` 
So `value.types[0]` is technically `[{argumentTypes: [Type, Type], resultType: Type}]` which isn't a Type object so `printType()` function won't work

I thought of your suggestion about changing the above to Type objects, but I realized it doesn't really make sense. E.g. like when we declare the type for BinaryExpression with operator `*` in TypeEnv, its types isn't related to a functionType although the syntax looks similar. Like I feel we shouldn't make it into something it isn't. So I think the current TypeEnv syntax should stay for the time being

Reply:
If I'm not wrong, aren't binary and unary operators functions? They take in arguments and return a result.

Reply:
Hmm okay makes sense. Then I'll refractor the TypeEnv ba. In that case you can push urs, it should work fine after my refactor